server/src/main/java/io/crate/collections/accountable/AccountableList.java

```diff
@@ -0,0 +1,189 @@
+ */
+public class AccountableList<T> extends AbstractList<T> {
+
+    private static final int SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private final LongConsumer allocateBytes;
+    private Object[] elementData;
+    private int size;
+
+    public AccountableList(LongConsumer allocateBytes) {
+        this.allocateBytes = allocateBytes;
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA; // Empty now, will be accounted on the first growth.
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public T get(int index) {
+        Objects.checkIndex(index, size);
+        return (T) elementData[index];
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    /**
+     * Similar to {@link ArrayList#sort(Comparator)} but without modification count checks.
+     */
+    public void sort(Comparator<? super T> c) {
+        Arrays.sort((T[]) elementData, 0, size, c);
+    }
+
+    @Override
+    public boolean add(T t) {
+        if (size == elementData.length) {
+            elementData = grow(size + 1);
+        }
+        elementData[size] = t;
+        size++;
+        return true;
+    }
+
+    /**
+     *
+     * @param fromIndex low endpoint (inclusive) of the subList
+     * @param toIndex high endpoint (exclusive) of the subList
+     * @return
+     */
+    public List<T> subList(int fromIndex, int toIndex) {
+        return new SubList<>(this, fromIndex, toIndex);
+    }
+
+    private static class SubList<T> extends AbstractList<T> implements RandomAccess {
+        private final AccountableList<T> root;
+        private final int offset;
+        private int size;
+
+        public SubList(AccountableList<T> root, int fromIndex, int toIndex) {
+            this.root = root;
+            root.allocateBytes.accept(NUM_BYTES_OBJECT_REF);
+            this.offset = fromIndex;
+            this.size = toIndex - fromIndex;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public void sort(Comparator<? super T> c) {
+            Arrays.sort((T[]) root.elementData, offset, offset + size, c);
+        }
+
+        @SuppressWarnings("unchecked")
+        public T set(int index, T element) {
+            Objects.checkIndex(index, size);
+            T oldValue = (T) root.elementData[offset + index];
+            root.elementData[offset + index] = element;
+            return oldValue;
+        }
+
+        @SuppressWarnings("unchecked")
+        public T get(int index) {
+            Objects.checkIndex(index, size);
+            return (T) root.elementData[offset + index];
+        }
+
+        public int size() {
+            return size;
+        }
+
+        public boolean add(T element) {
+            throw new UnsupportedOperationException();
+        }
```

Comment:
Can't this be removed as `AbstractList.add()` will throw the same exception already?

Reply:
Right, I saw that `AbstractList.add()` has some implementation and didn't notice that it calls method which throws an exception.